Kronos-Shifts-Connector/Microsoft.Teams.Shifts.Integration/Microsoft.Teams.Shifts.Integration.API/Controllers/TeamsController.cs

```diff
@@ -319,70 +325,85 @@ private static async Task<RequestModel> DecryptEncryptedRequestFromShiftsAsync(b
         /// <param name="updateProps">telemetry properties.</param>
         /// <returns>Returns list of ShiftIntegResponse for request.</returns>
-        private async Task<List<ShiftsIntegResponse>> ProcessOpenShiftRequest(RequestModel jsonModel, Dictionary<string, string> updateProps)
+        private async Task<List<ShiftsIntegResponse>> ProcessOpenShiftRequest(RequestModel jsonModel, Dictionary<string, string> updateProps, string teamsId, string workforceIntegrationId, string passThroughHeaderValue)
         {
             List<ShiftsIntegResponse> responseModelList = new List<ShiftsIntegResponse>();
             var requestBody = jsonModel.Requests.First(x => x.Url.Contains("/openshiftrequests/", StringComparison.InvariantCulture)).Body;
-            var requestState = requestBody != null ? requestBody["state"].Value<string>() : null;
 
             if (requestBody != null)
             {
+                var requestState = requestBody?["state"].Value<string>();
+
                 switch (requestState)
                 {
                     // The Open shift request is submitted in Shifts and is pending with manager for approval.
                     case ApiConstants.ShiftsPending:
                         {
-                            responseModelList = await this.ProcessOutboundOpenShiftRequestAsync(jsonModel, updateProps).ConfigureAwait(false);
+                            var openShiftRequest = JsonConvert.DeserializeObject<OpenShiftRequestIS>(requestBody.ToString());
+                            responseModelList = await this.ProcessOutboundOpenShiftRequestAsync(openShiftRequest, updateProps, teamsId).ConfigureAwait(false);
                         }
 
                         break;
 
                     // The Open shift request is approved by manager.
                     case ApiConstants.ShiftsApproved:
                         {
-                            responseModelList = await this.ProcessOpenShiftRequestApprovalAsync(jsonModel, updateProps).ConfigureAwait(false);
+                            // Passthrough header not present for outbound call i.e the request is coming from either Shifts or some unknown source.
+                            if (string.IsNullOrEmpty(passThroughHeaderValue))
+                            {
+                                this.telemetryClient.TrackTrace(Resource.PassthroughHeaderNotPresent);
+                                responseModelList = this.GenerateResponseToPreventAction(jsonModel, Resource.PreventManageActionFromShifts);
+                            }
+
+                            // Manager has approved open shift request in Kronos.
+                            else if (passThroughHeaderValue.Equals(workforceIntegrationId, StringComparison.Ordinal))
+                            {
+                                this.telemetryClient.TrackTrace(Resource.PassthroughHeaderWFIIDMatched);
```

Comment:
What is this for? If this request came from you, there should be no further processing required, no?

Reply:
@aosolis , This is the case when an open shift request is approved in shifts and we made a graph call to open shift request approve API as we have to update our DB records based on the incoming payload for system declined requests and respond to all the requests in incoming payload.

Reply:
(Can you please not resolve comments until you and I agree that it is resolved, as I can't unresolve them.)

I'm not following your logic. You're saying
1. Request was approved in Shifts. That results in a call to this controller, w/o the passthrough header.
2. That handler now makes updates to Shifts in Graph, and includes the passthrough header.
3. That results in another call to this controller, this time with your passthrough header.

Why is there more processing needed in Step 3? Shouldn't all the necessary work be done by the code that handles the request in Step 2?

Reply:
@aosolis , 
Below is the flow,
1. Request is approved in Kronos and as part of sync we fetched approved requests from 
    Kronos and made graph calls to approve the same request in Shifts which includes the 
    passthrough header.
2. While approving in Shifts (with Graph call), we get an outbound call, where we get the 
    passthrough header which we have passed in step 1. 
    We get below information from payload of outbound call: 
   a. The newly created shift as part of OpenShiftRequest approval, which we save in DB to 
       make shift entity up to date.
   b. The OpenShiftRequest which are declined by system while approving the current request,
       we update the status of such OpenShiftRequest(s) in our DB.